builder/dockerfile/internals.go

```diff
@@ -0,0 +1,658 @@
+	var origPaths string
+
+	if len(infos) == 1 {
+		fi := infos[0].FileInfo
+		origPaths = fi.Name()
+		if hfi, ok := fi.(builder.Hashed); ok {
+			srcHash = hfi.Hash()
+		}
+	} else {
+		var hashs []string
+		var origs []string
+		for _, info := range infos {
+			fi := info.FileInfo
+			origs = append(origs, fi.Name())
+			if hfi, ok := fi.(builder.Hashed); ok {
+				hashs = append(hashs, hfi.Hash())
+			}
+		}
+		hasher := sha256.New()
+		hasher.Write([]byte(strings.Join(hashs, ",")))
+		srcHash = "multi:" + hex.EncodeToString(hasher.Sum(nil))
+		origPaths = strings.Join(origs, " ")
+	}
+
+	cmd := b.runConfig.Cmd
+	if runtime.GOOS != "windows" {
+		b.runConfig.Cmd = stringutils.NewStrSlice("/bin/sh", "-c", fmt.Sprintf("#(nop) %s %s in %s", cmdName, srcHash, dest))
+	} else {
+		b.runConfig.Cmd = stringutils.NewStrSlice("cmd", "/S /C", fmt.Sprintf("REM (nop) %s %s in %s", cmdName, srcHash, dest))
+	}
+	defer func(cmd *stringutils.StrSlice) { b.runConfig.Cmd = cmd }(cmd)
+
+	if hit, err := b.probeCache(); err != nil {
+		return err
+	} else if hit {
+		return nil
+	}
+
+	container, _, err := b.docker.Create(b.runConfig, nil)
+	if err != nil {
+		return err
+	}
+	defer container.Unmount()
+	b.tmpContainers[container.ID] = struct{}{}
+
+	comment := fmt.Sprintf("%s %s in %s", cmdName, origPaths, dest)
+
+	// Twiddle the destination when its a relative path - meaning, make it
+	// relative to the WORKINGDIR
+	if !system.IsAbs(dest) {
+		hasSlash := strings.HasSuffix(dest, string(os.PathSeparator))
+		dest = filepath.Join(string(os.PathSeparator), filepath.FromSlash(b.runConfig.WorkingDir), dest)
+
+		// Make sure we preserve any trailing slash
+		if hasSlash {
+			dest += string(os.PathSeparator)
+		}
+	}
+
+	for _, info := range infos {
+		if err := b.docker.Copy(container, dest, info.FileInfo, info.decompress); err != nil {
+			return err
+		}
+	}
+
+	if err := b.commit(container.ID, cmd, comment); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
+	// get filename from URL
+	u, err := url.Parse(srcURL)
+	if err != nil {
+		return
+	}
+	path := u.Path
+	if strings.HasSuffix(path, string(os.PathSeparator)) {
+		path = path[:len(path)-1]
+	}
+	parts := strings.Split(path, string(os.PathSeparator))
+	filename := parts[len(parts)-1]
+	if filename == "" {
+		err = fmt.Errorf("cannot determine filename from url: %s", u)
+		return
+	}
+
+	// Initiate the download
+	resp, err := httputils.Download(srcURL)
+	if err != nil {
+		return
+	}
+
+	// Prepare file in a tmp dir
+	tmpDir, err := ioutils.TempDir("", "docker-remote")
+	if err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			os.RemoveAll(tmpDir)
+		}
+	}()
+	tmpFileName := filepath.Join(tmpDir, filename)
+	tmpFile, err := os.OpenFile(tmpFileName, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)
+	if err != nil {
+		return
+	}
+
+	// Download and dump result to tmp file
+	// TODO: put a better progressreader back
```

Comment:
```diff
diff --git a/builder/dockerfile/internals.go b/builder/dockerfile/internals.go
index 2319cd6..850da09 100644
--- a/builder/dockerfile/internals.go
+++ b/builder/dockerfile/internals.go
@@ -29,6 +29,8 @@ import (
 	"github.com/docker/docker/pkg/httputils"
 	"github.com/docker/docker/pkg/ioutils"
 	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/progressreader"
+	"github.com/docker/docker/pkg/streamformatter"
 	"github.com/docker/docker/pkg/stringid"
 	"github.com/docker/docker/pkg/stringutils"
 	"github.com/docker/docker/pkg/system"
@@ -271,20 +273,19 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
 
 	// Download and dump result to tmp file
 	// TODO: put a better progressreader back
-	/*
-		if _, err := io.Copy(tmpFile, progressreader.New(progressreader.Config{
-			      In:       resp.Body,
-			      Out:      b.Stdout,
-			      Formatter: ...,
-			      Size:     resp.ContentLength,
-			      NewLines: true,
-			      ID:       "",
-			      Action:   "Downloading",
-	*/
-	if _, err = io.Copy(tmpFile, resp.Body); err != nil {
+	if _, err = io.Copy(tmpFile, progressreader.New(progressreader.Config{
+		In:        resp.Body,
+		Out:       b.Stdout,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      resp.ContentLength,
+		NewLines:  true,
+		ID:        "",
+		Action:    "Downloading",
+	})); err != nil {
 		tmpFile.Close()
 		return
 	}
+
 	fmt.Fprintln(b.Stdout)
 	// ignoring error because the file was already opened successfully
 	tmpFileSt, _ := tmpFile.Stat()
```